arangod/Cluster/ClusterInfo.cpp

```diff
@@ -4208,6 +4213,136 @@ Result ClusterInfo::finishModifyingAnalyzerCoordinator(
   return Result(TRI_ERROR_NO_ERROR);
 }
 
+void ClusterInfo::initMetricsState() {
+  auto rebootId = ServerState::instance()->getRebootId().value();
+  auto serverId = ServerState::instance()->getId();
+
+  VPackBuilder builderRebootId;
+  builderRebootId.add(VPackValue{rebootId});
+  VPackBuilder builderServerId;
+  builderServerId.add(VPackValue{serverId});
+  VPackBuilder builderVersion;
+  builderVersion.add(VPackValue{0});
+
+  AgencyWriteTransaction const write{
+      {{kMetricsRebootId, AgencyValueOperationType::SET,
+        builderRebootId.slice()},
+       {kMetricsServerId, AgencyValueOperationType::SET,
+        builderServerId.slice()},
+       {kMetricsVersion, AgencyValueOperationType::SET,
+        builderVersion.slice()}},
+      {{kMetricsRebootId, AgencyPrecondition::Type::EMPTY, true},
+       {kMetricsServerId, AgencyPrecondition::Type::EMPTY, true},
+       {kMetricsVersion, AgencyPrecondition::Type::EMPTY, true}}};
+  AgencyComm ac{_server};
+  while (!server().isStopping()) {
+    auto const r = ac.sendTransactionWithFailover(write);
+    if (r.successful() ||
+        r.httpCode() == rest::ResponseCode::PRECONDITION_FAILED) {
+      return;
+    }
+  }
+}
+
+ClusterInfo::MetricsState ClusterInfo::getMetricsState(bool wantLeader) {
+  auto& ac = _server.getFeature<ClusterFeature>().agencyCache();
+  auto [result, index] = ac.read({AgencyCommHelper::path(kMetricsServerId),
+                                  AgencyCommHelper::path(kMetricsRebootId),
+                                  AgencyCommHelper::path(kMetricsVersion)});
+  auto data = result->slice().at(0).get(std::initializer_list<std::string_view>{
+      AgencyCommHelper::path(), "Plan", "Metrics"});
+  auto leaderRebootId = data.get("RebootId").getNumber<uint64_t>();
+  auto leaderServerId = data.get("ServerId").stringView();
+  auto leaderVersion = data.get("Version").getNumber<uint64_t>();
+  auto ourRebootId = ServerState::instance()->getRebootId().value();
+  auto ourServerId = ServerState::instance()->getId();
+  if (ourRebootId == leaderRebootId && ourServerId == leaderServerId) {
+    TRI_ASSERT(_metricsGuard.empty());
+    return {std::nullopt, leaderVersion};
+  }
+  if (wantLeader) {
+    // remove old callback (with _metricsGuard call, then store new callback)
+    _metricsGuard = _rebootTracker.callMeOnChange(
+        leaderServerId, RebootId{leaderRebootId},
+        [this, leaderRebootId, serverId = std::string{leaderServerId}] {
+          proposeMetricsLeader(leaderRebootId, serverId);
```

Comment:
So if the leader dies, in the worst case, we will have n (number of coordinators) * count of retry (mostly one) write transactions (only one will be successful)